### ISSUE TITLE
Touchscreen coordinates are rotated 90 degrees counter-clockwise from landscape user orientation

### ISSUE THREAD

**Issue by ThuggishNuggets** [stockmind/gpd-pocket-ubuntu-respin#51]
This is a continuation of a thread started on the GPD Pocket Owners Group on Facebook and is being recorded for posterity:
https://www.facebook.com/groups/144602682766328/permalink/171022066791056/?comment_id=171028816790381&reply_comment_id=171277686765494

The touchscreen issue occurred after upgrading the kernel:
``` 
Start: 4.12.0-rc2+ #261 // note: this is base GPD
  End: 4.13.0-rc5-custom #3 SMP Fri Aug 25 07:14:08 CEST 2017 x86_64 x86_64 x86_64 GNU/Linux
```
The kernel update fixes fast charging so that it no longer takes upwards of 12 hours to charge to full.

The kernel was updated using the following:
```
wget https://raw.githubusercontent.com/stockmind/gpd-pocket-ubuntu-respin/dev/update-kernel.sh
chmod +x update-kernel.sh
sudo ./update-kernel.sh
```
After updating the kernel, I ran the latest update script:
https://github.com/stockmind/gpd-pocket-ubuntu-respin/blob/master/update.sh

So now we're trying to resolve the rotated/unrotated touchscreen.

Here is the property list for the the `Goodix Capacitive TouchScreen id=11`:
```
$ xinput
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ SINO WEALTH Gaming Keyboard             	id=10	[slave  pointer  (2)]
⎜   ↳ Goodix Capacitive TouchScreen           	id=11	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ Power Button                            	id=6	[slave  keyboard (3)]
    ↳ Video Bus                               	id=7	[slave  keyboard (3)]
    ↳ Power Button                            	id=8	[slave  keyboard (3)]
    ↳ SINO WEALTH Gaming Keyboard             	id=9	[slave  keyboard (3)]
    ↳ Intel HID events                        	id=12	[slave  keyboard (3)]
    ↳ gpio-keys                               	id=13	[slave  keyboard (3)]
    ↳ gpio-keys                               	id=14	[slave  keyboard (3)]
```
```
$ xinput list-props 11
Device 'Goodix Capacitive TouchScreen':
	Device Enabled (140):	1
	Coordinate Transformation Matrix (142):	0.000000, 1.000000, 0.000000, -1.000000, 0.000000, 1.000000, 0.000000, 0.000000, 1.000000
	Device Accel Profile (265):	0
	Device Accel Constant Deceleration (266):	1.000000
	Device Accel Adaptive Deceleration (267):	1.000000
	Device Accel Velocity Scaling (268):	10.000000
	Device Product ID (258):	1046, 911
	Device Node (259):	"/dev/input/event6"
	Evdev Axis Inversion (269):	1, 0
	Evdev Axis Calibration (270):	<no items>
	Evdev Axes Swap (271):	1
	Axis Labels (272):	"Abs MT Position X" (291), "Abs MT Position Y" (292), "Abs MT Touch Major" (289), "Abs MT Width Major" (290), "None" (0), "None" (0)
	Button Labels (273):	"Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Wheel Up" (146), "Button Wheel Down" (147)
	Evdev Scrolling Distance (274):	0, 0, 0
	Evdev Middle Button Emulation (275):	0
	Evdev Middle Button Timeout (276):	50
	Evdev Third Button Emulation (277):	0
	Evdev Third Button Emulation Timeout (278):	1000
	Evdev Third Button Emulation Button (279):	3
	Evdev Third Button Emulation Threshold (280):	20
	Evdev Wheel Emulation (281):	0
	Evdev Wheel Emulation Axes (282):	0, 0, 4, 5
	Evdev Wheel Emulation Inertia (283):	10
	Evdev Wheel Emulation Timeout (284):	200
	Evdev Wheel Emulation Button (285):	4
	Evdev Drag Lock Buttons (286):	0
```
From what I can tell, still being very new to Linux myself, the problem is that the `Coordinate Transformation Matrix` property is set incorrectly relative to the user orientation of landscape.

I think that means this will fix the issue:
```
xinput set-prop 11 --type=float "Coordinate Transformation Matrix" 0 0 0 0 0 0 0 0 0
```
([Source](https://askubuntu.com/questions/368317/rotate-touch-input-with-touchscreen-and-or-touchpad))

**Comment by stockmind** [comment]
xinput set-prop 11 "Evdev Axes Swap" 0

     xinput set-prop 11 "Evdev Axis Inversion" 0 0

Issue first command then try touchscreen.
Then try second command and test touchscreen.
Keep transformation matrix as it is for this test.

**Comment by ThuggishNuggets** [comment]
Yeah the first command (setting Coordinate Transformation Matrix) sets it to landscape but the vertical axis is flipped and so it is landscape upside down.

**Comment by stockmind** [comment]
After both commands issued touchscreen axis are ok?
Commands are lost after reboot, so must be tested both in the same session without reboot.

**Comment by stockmind** [comment]
Also, do you mind to burn "gpdpocket-20170819-ubuntu-17.04-desktop-amd64.iso
" iso on a USB and test if on live environment your touch is ok?

You can download it from here:
https://mega.nz/#F!8WpQRZrD!0XHgajeG-QVZTp1Jbjndgw

**Comment by ThuggishNuggets** [comment]
No the axis are not functional. Their direction and axis appear flipped. I am now back to to upside down landscape using:
```
xinput set-prop 11 --type=float "Coordinate Transformation Matrix" 0 1 0 -1 0 1 0 0 1
xinput set-prop 11 "Evdev Axis Inversion" 1 0
xinput set-prop 11 "Evdev Axes Swap" 0
```

**Comment by ThuggishNuggets** [comment]
Fixed with the following:
```
xinput set-prop 11 --type=float "Coordinate Transformation Matrix" 0 1 0 -1 0 1 0 0 1
xinput set-prop 11 "Evdev Axis Inversion" 0 0
xinput set-prop 11 "Evdev Axes Swap" 0
```
Looks like its the same commands given by @stockmind but they didn't fix it the first time because I had already changed the [Coordinate Transformation Matrix (CTM)](https://wiki.ubuntu.com/X/InputCoordinateTransformation) to be all zeros. After resetting the CTM, setting the axis swap to `0`, and setting both axis inversions to `0`, it is fully functional. Woot!

**Comment by ThuggishNuggets** [comment]
## Working `xinput` touch screen configuration for kernel `4.13.0-rc5-custom #3 SMP Fri Aug 25 07:14:08 CEST 2017 x86_64 x86_64 x86_64 GNU/Linux`
```
$ xinput list-props 11
Device 'Goodix Capacitive TouchScreen':
	Device Enabled (140):	1
	Coordinate Transformation Matrix (142):	0.000000, 1.000000, 0.000000, -1.000000, 0.000000, 1.000000, 0.000000, 0.000000, 1.000000
	Device Accel Profile (265):	0
	Device Accel Constant Deceleration (266):	1.000000
	Device Accel Adaptive Deceleration (267):	1.000000
	Device Accel Velocity Scaling (268):	10.000000
	Device Product ID (258):	1046, 911
	Device Node (259):	"/dev/input/event6"
	Evdev Axis Inversion (269):	0, 0
	Evdev Axis Calibration (270):	<no items>
	Evdev Axes Swap (271):	0
	Axis Labels (272):	"Abs MT Position X" (291), "Abs MT Position Y" (292), "Abs MT Touch Major" (289), "Abs MT Width Major" (290), "None" (0), "None" (0)
	Button Labels (273):	"Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Wheel Up" (146), "Button Wheel Down" (147)
	Evdev Scrolling Distance (274):	0, 0, 0
	Evdev Middle Button Emulation (275):	0
	Evdev Middle Button Timeout (276):	50
	Evdev Third Button Emulation (277):	0
	Evdev Third Button Emulation Timeout (278):	1000
	Evdev Third Button Emulation Button (279):	3
	Evdev Third Button Emulation Threshold (280):	20
	Evdev Wheel Emulation (281):	0
	Evdev Wheel Emulation Axes (282):	0, 0, 4, 5
	Evdev Wheel Emulation Inertia (283):	10
	Evdev Wheel Emulation Timeout (284):	200
	Evdev Wheel Emulation Button (285):	4
	Evdev Drag Lock Buttons (286):	0
```

**Comment by ThuggishNuggets** [comment]
@stockmind [wrote](https://github.com/stockmind/gpd-pocket-ubuntu-respin/issues/51#issuecomment-326774360):
> Also, do you mind to burn `gpdpocket-20170819-ubuntu-17.04-desktop-amd64.iso` ... on a USB and test if on live environment your touch is ok?
> 
> You can download it from here:
> https://mega.nz/#F!8WpQRZrD!0XHgajeG-QVZTp1Jbjndgw

Yeah, I'll start the burn now and check it out when I get back on later tonight.

**Comment by stockmind** [comment]
pull latest changes from master branch and run update script.
i will probably drop support for gpd official kernel and investigate changes on latest ubuntu firmware

**Comment by stockmind** [comment]
Also when you boot from live usb, issue the same command to check props of touchscreen and report! Thanks

**Comment by ThuggishNuggets** [comment]
### Before running `update.sh`
Booted using live USB and everything appears to be working without any changes. 

### Kernel Info
```
ubuntu@ubuntu:~$ uname -ra
Linux ubuntu 4.13.0-rc3-custom-battery #2 SMP Sat Aug 12 06:42:35 CEST 2017 x86_64 x86_64 x86_64 GNU/Linux
```

### `xinput` output without updates
```
ubuntu@ubuntu:~$ xinput
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ Goodix Capacitive TouchScreen           	id=11	[slave  pointer  (2)]
⎜   ↳ SINO WEALTH Gaming Keyboard             	id=10	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ Power Button                            	id=6	[slave  keyboard (3)]
    ↳ Video Bus                               	id=7	[slave  keyboard (3)]
    ↳ Power Button                            	id=8	[slave  keyboard (3)]
    ↳ Intel HID events                        	id=12	[slave  keyboard (3)]
    ↳ Goodix Capacitive TouchScreen           	id=16	[slave  keyboard (3)]
    ↳ gpio-keys                               	id=14	[slave  keyboard (3)]
    ↳ gpio-keys                               	id=13	[slave  keyboard (3)]
    ↳ SINO WEALTH Gaming Keyboard             	id=9	[slave  keyboard (3)]
    ↳ SINO WEALTH Gaming Keyboard             	id=15	[slave  keyboard (3)]
```
```
ubuntu@ubuntu:~$ xinput list-props 11
Device 'Goodix Capacitive TouchScreen':
	Device Enabled (140):	1
	Coordinate Transformation Matrix (142):	0.000000, 1.000000, 0.000000, -1.000000, 0.000000, 1.000000, 0.000000, 0.000000, 1.000000
	libinput Calibration Matrix (295):	1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
	libinput Calibration Matrix Default (296):	1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
	libinput Send Events Modes Available (260):	1, 0
	libinput Send Events Mode Enabled (261):	0, 0
	libinput Send Events Mode Enabled Default (262):	0, 0
	Device Node (263):	"/dev/input/event6"
	Device Product ID (264):	1046, 911
```
Going to clone `master` and run `update.sh` then report back my findings.

**Comment by ThuggishNuggets** [comment]
Well I ran the update script and then rebooted and it was reset. Derp. Will do it again and report back. I won't be able to reboot since it will lose the changes though.

Side note: Really liking 17.04

**Comment by ThuggishNuggets** [comment]
### After running `update.sh`
### Kernel Info
```
$ uname -ra
Linux ubuntu 4.13.0-rc3-custom-battery #2 SMP Sat Aug 12 06:42:35 CEST 2017 x86_64 x86_64 x86_64 GNU/Linux
```
### `xinput` output after updates
```
ubuntu@ubuntu:~$ xinput
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ Goodix Capacitive TouchScreen           	id=11	[slave  pointer  (2)]
⎜   ↳ SINO WEALTH Gaming Keyboard             	id=10	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ Power Button                            	id=6	[slave  keyboard (3)]
    ↳ Video Bus                               	id=7	[slave  keyboard (3)]
    ↳ Power Button                            	id=8	[slave  keyboard (3)]
    ↳ Intel HID events                        	id=12	[slave  keyboard (3)]
    ↳ Goodix Capacitive TouchScreen           	id=16	[slave  keyboard (3)]
    ↳ gpio-keys                               	id=14	[slave  keyboard (3)]
    ↳ gpio-keys                               	id=13	[slave  keyboard (3)]
    ↳ SINO WEALTH Gaming Keyboard             	id=15	[slave  keyboard (3)]
    ↳ SINO WEALTH Gaming Keyboard             	id=9	[slave  keyboard (3)]
```
```
ubuntu@ubuntu:~$ xinput list-props 11
Device 'Goodix Capacitive TouchScreen':
	Device Enabled (140):	1
	Coordinate Transformation Matrix (142):	0.000000, 1.000000, 0.000000, -1.000000, 0.000000, 1.000000, 0.000000, 0.000000, 1.000000
	libinput Calibration Matrix (295):	1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
	libinput Calibration Matrix Default (296):	1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
	libinput Send Events Modes Available (260):	1, 0
	libinput Send Events Mode Enabled (261):	0, 0
	libinput Send Events Mode Enabled Default (262):	0, 0
	Device Node (263):	"/dev/input/event6"
	Device Product ID (264):	1046, 911
```
Note: I did not run `update-kernel.sh`

**Comment by ThuggishNuggets** [comment]
@stockmind the info you requested has been added.

The `gpdpocket-20170819-ubuntu-17.04-desktop-amd64.iso` image seems to be working very well even before running `update.sh`. I'm seriously considering just jumping on this version. Have you been seeing any problems with it or know of any? (never mind, I'll wait until #47 is resolved)

Built-in microphone not working with 17.04 iso, but then mine hasn't worked since I got it.

**Comment by stockmind** [comment]
I will release an update image with kernel rc5 that's the best at the moment. You can still install the image you already have and simply run update-kernel script to get latest kernel. Everything will be fine and exactly like an updated image.

**Comment by ThuggishNuggets** [comment]
Another problem on 4.13.0-rc3 is the brightness keys interact with the OS to set brightness, but the monitor itself doesn't have its brightness values modified. We can open a separate issue for that, but wanted to note it here so that we know that not everything is working on 17.04 with 4.13.0 kernel.

Maybe a Known Issues markdown in the root directory would be a good idea so we can contrast and compare the kernel and distro versions?

**Comment by stockmind** [comment]
Screen brightness works only after install at the moment.
That is warned in "What works" section. Could be added to https://github.com/stockmind/gpd-pocket-ubuntu-respin#troubleshooting
